language/internal/cc/parser/parser.go

```diff
@@ -25,28 +25,45 @@ package parser
 import (
 	"errors"
 	"fmt"
-	"io"
 	"log"
 	"os"
 	"regexp"
+	"slices"
 	"strconv"
 	"strings"
+
+	"github.com/EngFlow/gazelle_cc/internal/collections"
+	"github.com/EngFlow/gazelle_cc/language/internal/cc/lexer"
 )
 
-// ParseSource runs the extractor on an in‑memory buffer.
-func ParseSource(input string) (SourceInfo, error) {
-	return parse(strings.NewReader(input))
+func isRelevantTokenType(token lexer.Token) bool {
+	switch token.Type {
+	case lexer.TokenType_Word, lexer.TokenType_Symbol, lexer.TokenType_PreprocessorDirective, lexer.TokenType_Newline:
+		return true
+	default:
+		return false
+	}
+}
+
+// ParseSource reads and parses C/C++ source, returning structured SourceInfo.
+func ParseSource(input []byte) (SourceInfo, error) {
+	allTokens := lexer.NewLexer(input).Tokenize()
+	filteredTokens := collections.Filter(allTokens, isRelevantTokenType)
```

Comment:
We're creating a big slice of tokens with the `.Tokenize` call, then copying and filtering to another slice. That seems a little wasteful of memory.

Instead, I think it's better to call `NextToken()` repeatedly in a loop and append to `filteredTokens` if it's one of the relevant types.

Optional, but if you want to go back to a method that returns an `iter.Seq`, we could loop over that directly. I think you had it earlier, but it required a separate error check. Now that `NextToken` doesn't return an error, it should work better.